File: benchmarks/crc32/c/crc32.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

static uint32_t crc32_table[256];

static void build_table(void) {
    for (uint32_t i = 0; i < 256; i++) {
        uint32_t crc = i;
        for (int j = 0; j < 8; j++) {
            if (crc & 1) crc = (crc >> 1) ^ 0xEDB88320u;
            else crc >>= 1;
        }
        crc32_table[i] = crc;
    }
}

static uint32_t crc32_compute(const unsigned char *data, size_t len) {
    uint32_t crc = 0xFFFFFFFFu;
    for (size_t i = 0; i < len; i++) {
        crc = crc32_table[(crc ^ data[i]) & 0xFF] ^ (crc >> 8);
    }
    return crc ^ 0xFFFFFFFFu;
}

int main(int argc, char **argv) {
    long n = argc > 1 ? atol(argv[1]) : 50000000;

    build_table();

    if (crc32_compute((const unsigned char *)"", 0) != 0x00000000u) {
        fprintf(stderr, "self-check failed: CRC32(\"\") mismatch\n");
        return 1;
    }
    const unsigned char check[] = "123456789";
    if (crc32_compute(check, 9) != 0xCBF43926u) {
        fprintf(stderr, "self-check failed: CRC32(\"123456789\") mismatch\n");
        return 1;
    }

    unsigned char *buf = malloc((size_t)n);
    for (long i = 0; i < n; i++) buf[i] = (unsigned char)((i * 131 + 7) % 256);

    uint32_t result = crc32_compute(buf, (size_t)n);
    printf("%08x\n", result);

    free(buf);
    return 0;
}

File: benchmarks/crc32/rust/crc32.rs
use std::env;

fn build_table() -> [u32; 256] {
    let mut table = [0u32; 256];
    for i in 0..256u32 {
        let mut crc = i;
        for _ in 0..8 {
            if crc & 1 != 0 {
                crc = (crc >> 1) ^ 0xEDB88320;
            } else {
                crc >>= 1;
            }
        }
        table[i as usize] = crc;
    }
    table
}

fn crc32_compute(table: &[u32; 256], data: &[u8]) -> u32 {
    let mut crc: u32 = 0xFFFFFFFF;
    for &b in data {
        crc = table[((crc ^ b as u32) & 0xFF) as usize] ^ (crc >> 8);
    }
    crc ^ 0xFFFFFFFF
}

fn main() {
    let n: usize = env::args().nth(1).map(|s| s.parse().unwrap()).unwrap_or(50_000_000);

    let table = build_table();

    assert_eq!(crc32_compute(&table, b""), 0x00000000, "self-check failed: CRC32(\"\") mismatch");
    assert_eq!(crc32_compute(&table, b"123456789"), 0xCBF43926, "self-check failed: CRC32(\"123456789\") mismatch");

    let mut buf = vec![0u8; n];
    for i in 0..n {
        buf[i] = ((i * 131 + 7) % 256) as u8;
    }

    let result = crc32_compute(&table, &buf);
    println!("{:08x}", result);
}
